### PR TITLE
refactor(scheduler): move scheduler to new crate to support workflows in multiple crates

### PR DESCRIPTION
**Pull request by jagan-jaya**
## Type of Change

- [ ] Bugfix
- [ ] New feature
- [X] Enhancement
- [X] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

Currently process tracker is tightly coupled with router that make it impossible for other crate/services to write its own workflow within their crates. So moving process tracker to new standalone crate `scheduler` so that a better interface can be enforced and external services can also use it as a service.

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## Motivation and Context

Writing individual workflows for external crates is impossible now as it requires all the workflows to be created in the router crate as it is tightly coupled. We are moving all the scheduler related changes to a new crate and exposing scheduler as a service so that external services can use scheduler as a library

This involves changes in three crates mainly
1. router
2. storage_models
3. scheduler

## Checklist

- [ ] I formatted the code `cargo +nightly fmt --all`
- [ ] I addressed lints thrown by `cargo clippy`
- [ ] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/router/src/routes/app.rs [hunk 1]
```diff
@@ -27,17 +28,23 @@ pub struct AppState {
...
+    }
+}
+
 pub trait AppStateInfo {
     fn conf(&self) -> settings::Settings;
     fn flow_name(&self) -> String;
     fn store(&self) -> Box<dyn StorageInterface>;
     #[cfg(feature = "email")]
-    fn email_client(&self) -> Box<dyn EmailClient>;
+    fn email_client(&self) -> Option<Box<dyn EmailClient>>;
```
Comment (NishantJoshi00) [inline crates/router/src/routes/app.rs:47]: Why is the made optional?

File: crates/scheduler/src/consumer.rs [hunk 2]
```diff
@@ -34,10 +30,10 @@ pub fn valid_business_statuses() -> Vec<&'static str> {
 }
 
 #[instrument(skip_all)]
-pub async fn start_consumer(
-    state: &AppState,
-    settings: sync::Arc<settings::SchedulerSettings>,
-    workflow_selector: workflows::WorkflowSelectorFn,
+pub async fn start_consumer<T: SchedulerAppState + Send + Sync + Clone + 'static>(
```
Comment (NishantJoshi00) [inline crates/scheduler/src/consumer.rs:33]: I think the `state` won't have `'static` lifetime. As the construction will happen at runtime. Still we might need to test this out

File: crates/scheduler/src/consumer.rs [hunk 3]
```diff
@@ -195,21 +195,25 @@ pub async fn fetch_consumer_tasks(
...
     process: storage::ProcessTracker,
-    operation: Box<dyn ProcessTrackerWorkflow>,
-) {
+    operation: Box<dyn ProcessTrackerWorkflow<T>>,
+) where
+    T: SchedulerAppState + Send + Sync + Clone + 'static,
+{
     let output = operation.execute_workflow(state, process.clone()).await;
     match output {
         Ok(_) => operation.success_handler(state, process).await,
```
Comment (NishantJoshi00) [inline crates/scheduler/src/consumer.rs:219]: we plan on moving this inside the workflow handler, right?

File: crates/scheduler/src/utils.rs [hunk 4]
```diff
@@ -243,15 +244,16 @@ pub fn get_time_from_delta(delta: Option<i32>) -> Option<time::PrimitiveDateTime
     delta.map(|t| common_utils::date_time::now().saturating_add(time::Duration::seconds(t.into())))
 }
 
-pub async fn consumer_operation_handler<E>(
-    state: AppState,
+pub async fn consumer_operation_handler<E, T: Send + Sync + 'static>(
```
Comment (NishantJoshi00) [inline crates/scheduler/src/utils.rs:247]: This should also have `AppState*` as trait bound?

**Review by NishantJoshi00 — CHANGES_REQUESTED** [review]

**Comment by ashokkjag** [comment]
At a high level, changes look OK. the diesel model crate seems to have config defaults/validation which is unexpected.

File: crates/scheduler/src/utils.rs (hunk 4 above)
Comment (jagan-jaya) [reply crates/scheduler/src/utils.rs:247]: Changed it to `SchedulerAppState`

File: crates/storage_impl/src/lib.rs [hunk 5]
```diff
@@ -222,6 +283,79 @@ impl DataModelExt for data_models::MerchantStorageScheme {
...
+/// Currently using `StrongSecret` won't have any effect as this struct have smart pointers to heap
+/// allocations.
+/// note: we can consider adding such behaviour in the future with custom implementation
+#[derive(Clone)]
+pub struct ActiveKmsSecrets {
+    pub jwekey: masking::Secret<Jwekey>,
+}
+#[derive(Debug, Deserialize, Clone, Default)]
+#[serde(default)]
+pub struct Jwekey {
```
Comment (lsampras) [inline crates/storage_impl/src/lib.rs:305]: does this need to be part of storage_impl ?

we can move it to common_utils if it's also used in scheduler else we can retain it in router

**Review by lsampras — APPROVED** [review]
rest looks good

File: crates/diesel_models/src/helpers.rs
Comment (jarnura) [inline crates/diesel_models/src/helpers.rs:1]: these are api client related things, move this from diesel models to router or some common_utils.

**Review by jarnura — CHANGES_REQUESTED** [review]

File: crates/external_services/src/email.rs [hunk 6]
```diff
@@ -121,3 +125,12 @@ pub enum AwsSesError {
     #[error("Failed to Send Email {0:?}")]
     SendingFailure(aws_smithy_client::SdkError<SendEmailError>),
 }
+
```
Comment (NishantJoshi00) [inline crates/external_services/src/email.rs:128]: check this

**Review by NishantJoshi00 — CHANGES_REQUESTED** [review]
Other than that, looks good 👍

File: crates/router/src/core/errors/api_error_response.rs [hunk 7]
```diff
@@ -227,6 +228,12 @@ pub enum ApiErrorResponse {
     CurrencyNotSupported { message: String },
 }
 
+impl PTError for ApiErrorResponse {
+    fn to_pt_error(&self) -> ProcessTrackerError {
+        ProcessTrackerError::EClientError
+    }
+}
```
[the remark this answers is not in the document]
Comment (NishantJoshi00) [reply crates/router/src/core/errors/api_error_response.rs:235]:
```rust
impl PTError for ApiErrorResponse {
    fn to_pt_error(self) -> ProcessTrackerError {
        ProcessTrackerError::EApiErrorResponse(error_stack::report!(self))
    }
}
```

File: crates/router/Cargo.toml [hunk 8]
```diff
@@ -16,9 +16,9 @@ email = ["external_services/email", "dep:aws-config"]
 basilisk = ["kms"]
 stripe = ["dep:serde_qs"]
 release = ["kms", "stripe", "basilisk", "s3", "email","accounts_cache"]
-olap = ["data_models/olap", "storage_impl/olap"]
+olap = ["data_models/olap", "storage_impl/olap", "scheduler/olap"]
```
[the remark this answers is not in the document]
Comment (jagan-jaya) [reply crates/router/Cargo.toml:19]: Yes scheduler is a dependency of router, so need olap feature in router

File: crates/scheduler/Cargo.toml [hunk 9]
```diff
@@ -0,0 +1,59 @@
+[package]
+name = "scheduler"
+version = "0.1.0"
+edition = "2021"
+
+[features]
+default = ["kv_store", "olap"]
+olap = []
```
[the remark this answers is not in the document]
Comment (jagan-jaya) [reply crates/scheduler/Cargo.toml:8]: Required as olap feature is required if enabled in router

**Review by jarnura — APPROVED** [review]

**Review by NishantJoshi00 — APPROVED** [review]

**Review by lsampras — APPROVED** [review]

**Pull request merged**